server/portal/apps/auth/models.py

```diff
@@ -79,16 +81,17 @@ def client(self):
                        refresh_token=self.refresh_token)
 
         with transaction.atomic():
-            if self.expired:
-                try:
-                    client.refresh_tokens()
-                except Exception:
-                    logger.exception('Tapis Token refresh failed')
-                    raise
-
-                self.update(created=int(time.time()),
-                            access_token=client.access_token.access_token,
-                            expires_in=client.access_token.expires_in().total_seconds())
+            try:
+                with self.lock:
+                    if self.expired:
```

Comment:
I think `select_for_update` might need to be utilized here somehow for `expired`.  and the `self.update()`would need to be moved within the `with self.lock` block as well.  then the behavior is that we lock things and check that if its expired and if expired, we update and then we release the lock.

Reply:
This will aquire a lock even if the token is not expired. It is aggressive, but guarantees correctness (which is what we want).

Reply:
true. good point 🤔 .  if its too aggressive, could we only use lock/select_for_update after a first check of`self.expired`.  so basically, we check if we think it is expired via `self.expired`, then get a lock and re-read `expired` again via `select_for_update` to determine if it is really expired (and not that someone else has just updated it) before updating it?

Reply:
> self.update()would need to be moved within the with self.lock block as well.)

Good catch - that is an bug, this code is updating even if it is not expired, which will cause the token to never expire on our end. A second related point, refresh_token is already doing self.update(), we do not need to do this again here. 

> on checking self.expired twice (once before and once after locking):

We could do this double checked locking. 
if self.expired(): 
   with self.lock:
      if self.expired():


A side note, in many other instances we moved away from this double checked locking due to bad compiler and runtime optimizations on this pattern. Hence, aggressive locking is ok sometimes or a quick api call is even ok if it is redundant to avoid correctness issues.

Reply:
thanks all. I've updated this so that it checks expired both before and after the lock, and I removed that extra update call. let me know if you notice anything else that should be modified or if i misunderstood anything.